File: backend/game/services/field_diff.py
"""
Field diff utilities for undo functionality.
These functions compute and apply diffs between field states,
storing only the changed cells for memory efficiency.
"""

import copy


def compute_field_diff(old_field: list, new_field: list, width: int, height: int) -> list:
    """
    Compute the diff between two field states.
    Returns a list of cells that changed, storing the ORIGINAL values (from old_field).

    Args:
        old_field: Field state before the change
        new_field: Field state after the change
        width: Field width
        height: Field height

    Returns:
        List of {"x": x, "y": y, "cell": cell_data} dicts representing original cell states
    """
    diff = []

    for x in range(width):
        for y in range(height):
            old_cell = old_field[x][y]
            new_cell = new_field[x][y]

            if not cells_equal(old_cell, new_cell):
                # Store the original cell (deep copy to preserve state)
                diff.append(
                    {
                        "x": x,
                        "y": y,
                        "cell": copy.deepcopy(old_cell),
                    }
                )

    return diff


def cells_equal(cell_a: dict, cell_b: dict) -> bool:
    """
    Check if two cells are equal (same unit, building state).
    Note: We ignore isHidden as that's visibility, not game state.

    Args:
        cell_a: First cell to compare
        cell_b: Second cell to compare

    Returns:
        True if cells are equal
    """
    # Compare units
    unit_a = cell_a.get("unit")
    unit_b = cell_b.get("unit")

    if (unit_a is None) != (unit_b is None):
        return False
    if unit_a and unit_b:
        if unit_a.get("player") != unit_b.get("player"):
            return False
        if unit_a.get("movePoints") != unit_b.get("movePoints"):
            return False
        if unit_a.get("hasMoved") != unit_b.get("hasMoved"):
            return False

    # Compare buildings
    building_a = cell_a.get("building")
    building_b = cell_b.get("building")

    if (building_a is None) != (building_b is None):
        return False
    if building_a and building_b:
        if building_a.get("player") != building_b.get("player"):
            return False
        if building_a.get("_type") != building_b.get("_type"):
            return False

    return True


def apply_field_diff(field: list, diff: list) -> None:
    """
    Apply a diff to restore the field to a previous state.
    Modifies the field in place.

    Args:
        field: Current field to modify
        diff: List of {"x": x, "y": y, "cell": cell_data} dicts to restore
    """
    for item in diff:
        x = item["x"]
        y = item["y"]
        cell = item["cell"]

        # Restore the original cell (deep copy to avoid reference issues)
        field[x][y]["unit"] = copy.deepcopy(cell.get("unit"))
        field[x][y]["building"] = copy.deepcopy(cell.get("building"))

File: frontend/src/game/fieldDiff.js
/**
 * Field diff utilities for undo functionality.
 * These functions compute and apply diffs between field states,
 * storing only the changed cells for memory efficiency.
 */

import Models from '@/game/models'

/**
 * Compute the diff between two field states.
 * Returns an array of cells that changed, storing the ORIGINAL values (from oldField).
 *
 * @param {Array} oldField - Field state before the change
 * @param {Array} newField - Field state after the change
 * @param {number} width - Field width
 * @param {number} height - Field height
 * @returns {Array} Array of { x, y, cell } objects representing original cell states
 */
export function computeFieldDiff(oldField, newField, width, height) {
  const diff = []

  for (let x = 0; x < width; x++) {
    for (let y = 0; y < height; y++) {
      const oldCell = oldField[x][y]
      const newCell = newField[x][y]

      // Compare cells - check unit and building changes
      if (!cellsEqual(oldCell, newCell)) {
        // Store the original cell (deep clone to preserve state)
        // Use JSON.parse/stringify for compatibility with Vue reactive proxies
        diff.push({
          x,
          y,
          cell: JSON.parse(JSON.stringify(oldCell)),
        })
      }
    }
  }

  return diff
}

/**
 * Check if two cells are equal (same unit, building state).
 * Note: We ignore isHidden as that's visibility, not game state.
 *
 * @param {Object} cellA - First cell to compare
 * @param {Object} cellB - Second cell to compare
 * @returns {boolean} True if cells are equal
 */
function cellsEqual(cellA, cellB) {
  // Compare units
  const unitA = cellA.unit
  const unitB = cellB.unit

  if ((unitA === null || unitA === undefined) !== (unitB === null || unitB === undefined)) {
    return false
  }
  if (unitA && unitB) {
    if (unitA.player !== unitB.player) return false
    if (unitA.movePoints !== unitB.movePoints) return false
    if (unitA.hasMoved !== unitB.hasMoved) return false
  }

  // Compare buildings
  const buildingA = cellA.building
  const buildingB = cellB.building

  if (
    (buildingA === null || buildingA === undefined) !==
    (buildingB === null || buildingB === undefined)
  ) {
    return false
  }
  if (buildingA && buildingB) {
    if (buildingA.player !== buildingB.player) return false
    if (buildingA._type !== buildingB._type) return false
  }

  return true
}

/**
 * Apply a diff to restore the field to a previous state.
 * Modifies the field in place.
 *
 * @param {Array} field - Current field to modify
 * @param {Array} diff - Array of { x, y, cell } objects to restore
 */
export function applyFieldDiff(field, diff) {
  for (const { x, y, cell } of diff) {
    // Rebuild as Unit/Building instances — GameCell's prop validators
    // require `instanceof`, and the stored cell was deep-cloned to plain
    // objects to survive Vue reactive proxies. fromJSON also gives us a
    // fresh copy, so no reference leak back into the diff.
    field[x][y].unit = Models.Unit.fromJSON(cell.unit)
    field[x][y].building = Models.Building.fromJSON(cell.building)
  }
}
